### DISCUSSION TITLE
Implementation of subpixel smoothing

### DISCUSSION THREAD

**Discussion by lucasgrjn** [NanoComp/meep#2426] [Q&A]
Hello everyone !

I like to use Meep for my research and it is an increible tool ! But there is a little point which I don't understand the implementation, the subpixel smoothing part. From the dedicated [page](https://meep.readthedocs.io/en/latest/Subpixel_Smoothing/) and from the paper [1](https://opg.optica.org/ol/abstract.cfm?uri=ol-31-20-2972) and [2](https://opg.optica.org/ol/fulltext.cfm?uri=ol-34-18-2778&id=185824), I get the point we need to use a kind of projection matrix and the normal to interfaces from a theorical point of view. What I don't get is how we can get since our grid is discretized.

Basically how can implement the smoothing algorithm ? (I would like to understand to see if I can extend its use for some mode calculations tools we use)

Any help is welcome ! Thanks in advance,

Lucas

**Comment by smartalecH** [comment]
> What I don't get is how we can get since our grid is discretized.

The key point is that the geometry is first defined on a _continuous_ domain via a level set. The subpixel smoothing routine computes the _average_ permittivity within a voxel (via a smoothing kernel). 

The corresponding averaging integrals simplify quite a bit (with some approximations about how an interface cuts the voxel) with these level sets. These allow one to avoid performing ugly quadrature on some black box function that describes the geometry. 

For the fully general case, you'll have to perform these expensive quadratures. It gets a bit hairy with the yee grid. But effectively, this is how you deal with "the discrete grid" (you first work with a continuous domain). 

> (I would like to understand to see if I can extend its use for some mode calculations tools we use)

Are you just doing 2D mode calculations? If so, I wouldn't go through the hassle of second-order accurate smoothing. Just implement first-order accurate smoothing (a simple average, for example) and crank the resolution.

**Reply by lucasgrjn** [reply]
Actually we use the method, as you mentionned, based on an increase of the resolution followed by a simple average for 2D and 3D case. The main idea was evaluating the complexity to add a subpixel smoothing.

**Reply by smartalecH** [reply]
For 3D I could see it being _useful_ (that's what MPB does after all...)

Whether it's _worth it_ is another story. Software development time is usually more expensive than typical compute time (at least in the regime of payoffs possible here). 

More importantly, if you have to do brute force quadrature all over the place, it might end up making your _net_ computation _more expensive_ (depending on how you do it, what the problem looks like, etc).

In other words, I don't think it's a clear "win" to implement. But it could certainly be useful depending on your problem, resource constraints, etc.

**Reply by lucasgrjn** [reply]
I see your point and I fully agree with you :)

Just for my scientific curiosity I will try to see how it works in that case. So, basically you define the structure and then you transform it on a mathematical object via level-set methods. Then the normal vector of the interface can be calculated using the level-set shape and a bilinear interpolation. And Bingo!!! Now, you can find the smoother epsilon.

Am I right ?

**Reply by smartalecH** [reply]
> So, basically you define the structure and then you transform it on a mathematical object via level-set methods.

Im not quite sure I understand what you mean here. 

But basically, the level set helps simplify things for two reasons: (1) can compute an analytic normal direction (like you said); (2) it's easy to compute the fill fraction. 

These two quantities are used within the projection matrix etc that computes the final permittivity tensor. 

But you don't _have_ to use a level set. You just need a way to compute those quantities, and a level set makes that easy.

**Reply by lucasgrjn** [reply]
Thanks for the explanation! I think I get the point :)
